File: hushh-webapp/ios/App/App/Plugins/KaiPlugin.swift
import Capacitor
import Foundation

/**
 * Kai Plugin - iOS Implementation
 * 
 * Native plugin for Agent Kai stock analysis.
 * Makes HTTP calls to backend from native code.
 */

@objc(KaiPlugin)
public class KaiPlugin: CAPPlugin, CAPBridgedPlugin {
    
    // URLSession with 30s timeout
    private lazy var urlSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 30
        return URLSession(configuration: config)
    }()
    
    // MARK: - CAPBridgedPlugin Protocol
    public let identifier = "KaiPlugin"
    public let jsName = "Kai"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "grantConsent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "analyze", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storePreferences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPreferences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resetPreferences", returnType: CAPPluginReturnPromise)
    ]
    
    // MARK: - Configuration
    
    private var defaultBackendUrl: String {
        let configUrl = bridge?.config.getPluginConfig("Kai").getString("backendUrl")
        print("[KaiPlugin] 🥖 config.getPluginConfig(\"Kai\").getString(\"backendUrl\") check: \(configUrl ?? "nil")")
        return configUrl ?? "https://consent-protocol-1006304528804.us-central1.run.app"
    }

    private func getBackendUrl(_ call: CAPPluginCall) -> String {
        // 1. Check call parameters
        if let url = call.getString("backendUrl") {
            print("[KaiPlugin] 🌐 Using backendUrl from call params: \(url)")
            return url
        }

        // 2. Check capacitor config (Plugin specific: plugins.Kai.backendUrl)
        if let url = bridge?.config.getPluginConfig("Kai").getString("backendUrl") {
            print("[KaiPlugin] 🌐 Using backendUrl from Kai plugin config: \(url)")
            return url
        }

        // 3. Check for environment variable (as a fallback, though unlikely to be accessible here)
        if let envUrl = ProcessInfo.processInfo.environment["NEXT_PUBLIC_BACKEND_URL"] {
             print("[KaiPlugin] 🌐 Using backendUrl from Environment: \(envUrl)")
             return envUrl
        }

        // 4. Default
        let url = defaultBackendUrl
        print("[KaiPlugin] 🌐 Using final fallback backendUrl: \(url)")
        return url
    }
    
    // MARK: - Plugin Methods
    
    @objc func grantConsent(_ call: CAPPluginCall) {
        print("[KaiPlugin] 🔍 grantConsent called")
        guard let userId = call.getString("userId"),
              let scopes = call.getArray("scopes", String.self) else {
            print("[KaiPlugin] ❌ Missing required parameters: userId, scopes")
            call.reject("Missing required parameters: userId, scopes")
            return
        }
        
        let authToken = call.getString("authToken")
        let backendUrl = getBackendUrl(call)
        let urlStr = "\(backendUrl)/api/kai/consent/grant"
        print("[KaiPlugin] 🌐 URL: \(urlStr)")
        
        guard let url = URL(string: urlStr) else {
            call.reject("Invalid URL: \(urlStr)")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        let body: [String: Any] = [
            "user_id": userId,
            "scopes": scopes
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        urlSession.dataTask(with: request) { data, response, error in
            if let error = error {
                let errorMsg = "Network error: \(error.localizedDescription) | backendUrl: \(backendUrl)"
                print("[KaiPlugin] ❌ \(errorMsg)")
                call.reject(errorMsg)
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                print("[KaiPlugin] 📡 Response status: \(httpResponse.statusCode)")
                if !(200...299).contains(httpResponse.statusCode) {
                    let bodyStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "no body"
                    let truncatedBody = bodyStr.count > 200 ? String(bodyStr.prefix(200)) + "..." : bodyStr
                    let errorMsg = "HTTP Error \(httpResponse.statusCode) | backendUrl: \(backendUrl) | body: \(truncatedBody)"
                    print("[KaiPlugin] ❌ \(errorMsg)")
                    call.reject(errorMsg)
                    return
                }
            }
            
            guard let data = data else {
                print("[KaiPlugin] ❌ No data received")
                call.reject("No data received")
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    print("[KaiPlugin] ✅ grantConsent success: \(json.keys)")
                    call.resolve(json)
                } else if let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
                    print("[KaiPlugin] ✅ grantConsent success (Array)")
                    call.resolve(["data": array])
                } else {
                    print("[KaiPlugin] ❌ Invalid response format")
                    call.reject("Invalid response format")
                }
            } catch {
                print("[KaiPlugin] ❌ JSON parsing error: \(error.localizedDescription)")
                call.reject("JSON parsing error: \(error.localizedDescription)")
            }
        }.resume()
    }
    
    @objc func analyze(_ call: CAPPluginCall) {
        print("[KaiPlugin] 🔍 analyze called")
        guard let userId = call.getString("userId"),
              let ticker = call.getString("ticker"),
              let consentToken = call.getString("consentToken"),
              let riskProfile = call.getString("riskProfile"),
              let processingMode = call.getString("processingMode") else {
            print("[KaiPlugin] ❌ Missing required parameters")
            call.reject("Missing required parameters")
            return
        }
        
        let authToken = call.getString("authToken")
        let backendUrl = getBackendUrl(call)
        let urlStr = "\(backendUrl)/api/kai/analyze"
        print("[KaiPlugin] 🌐 URL: \(urlStr)")
        
        guard let url = URL(string: urlStr) else {
            call.reject("Invalid URL: \(urlStr)")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        var body: [String: Any] = [
            "user_id": userId,
            "ticker": ticker,
            "consent_token": consentToken,
            "risk_profile": riskProfile,
            "processing_mode": processingMode
        ]
        
        // Include context if provided
        if let context = call.getObject("context") {
            body["context"] = context
        }
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        urlSession.dataTask(with: request) { data, response, error in
            if let error = error {
                let errorMsg = "Network error: \(error.localizedDescription) | backendUrl: \(backendUrl)"
                print("[KaiPlugin] ❌ \(errorMsg)")
                call.reject(errorMsg)
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                print("[KaiPlugin] 📡 Response status: \(httpResponse.statusCode)")
                if !(200...299).contains(httpResponse.statusCode) {
                    let bodyStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "no body"
                    let truncatedBody = bodyStr.count > 200 ? String(bodyStr.prefix(200)) + "..." : bodyStr
                    let errorMsg = "HTTP Error \(httpResponse.statusCode) | backendUrl: \(backendUrl) | body: \(truncatedBody)"
                    print("[KaiPlugin] ❌ \(errorMsg)")
                    call.reject(errorMsg)
                    return
                }
            }
            
            guard let data = data else {
                print("[KaiPlugin] ❌ No data received")
                call.reject("No data received")
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    print("[KaiPlugin] ✅ analyze success")
                    // Aligned with Android: return flat JSON directly
                    call.resolve(json)
                } else {
                    print("[KaiPlugin] ❌ Invalid response format")
                    call.reject("Invalid response format")
                }
            } catch {
                print("[KaiPlugin] ❌ JSON parsing error: \(error.localizedDescription)")
                call.reject("JSON parsing error: \(error.localizedDescription)")
            }
        }.resume()
    }
    
    @objc func storePreferences(_ call: CAPPluginCall) {
        print("[KaiPlugin] 🔍 storePreferences called")
        guard let userId = call.getString("userId") else {
            print("[KaiPlugin] ❌ Missing required parameters")
            call.reject("Missing required parameters")
            return
        }

        // Canonical payload: preferences array (preferred)
        var preferencesPayload: Any? = nil
        if let preferences = call.getArray("preferences", JSObject.self) {
            preferencesPayload = preferences
        } else if let preferencesEncrypted = call.getString("preferencesEncrypted") {
            // Legacy payload: stringified JSON array
            if let data = preferencesEncrypted.data(using: .utf8) {
                preferencesPayload = try? JSONSerialization.jsonObject(with: data)
            }
        }

        guard preferencesPayload != nil else {
            call.reject("Missing preferences payload")
            return
        }
        
        let authToken = call.getString("authToken")
        let backendUrl = getBackendUrl(call)
        let urlStr = "\(backendUrl)/api/kai/preferences/store"
        print("[KaiPlugin] 🌐 URL: \(urlStr)")
        
        guard let url = URL(string: urlStr) else {
            call.reject("Invalid URL: \(urlStr)")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        let body: [String: Any] = [
            "user_id": userId,
            "preferences": preferencesPayload as Any
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        urlSession.dataTask(with: request) { data, response, error in
            if let error = error {
                let errorMsg = "Network error: \(error.localizedDescription) | backendUrl: \(backendUrl)"
                print("[KaiPlugin] ❌ \(errorMsg)")
                call.reject(errorMsg)
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                print("[KaiPlugin] 📡 Response status: \(httpResponse.statusCode)")
                if !(200...299).contains(httpResponse.statusCode) {
                    let bodyStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "no body"
                    let truncatedBody = bodyStr.count > 200 ? String(bodyStr.prefix(200)) + "..." : bodyStr
                    let errorMsg = "HTTP Error \(httpResponse.statusCode) | backendUrl: \(backendUrl) | body: \(truncatedBody)"
                    print("[KaiPlugin] ❌ \(errorMsg)")
                    call.reject(errorMsg)
                    return
                }
            }
            
            guard let data = data else {
                print("[KaiPlugin] ❌ No data received")
                call.reject("No data received")
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    print("[KaiPlugin] ✅ storePreferences success")
                    call.resolve(json)
                } else {
                    print("[KaiPlugin] ❌ Invalid response format")
                    call.reject("Invalid response format")
                }
            } catch {
                print("[KaiPlugin] ❌ JSON parsing error: \(error.localizedDescription)")
                call.reject("JSON parsing error: \(error.localizedDescription)")
            }
        }.resume()
    }
    
    @objc func getPreferences(_ call: CAPPluginCall) {
        print("[KaiPlugin] 🔍 getPreferences called")
        guard let userId = call.getString("userId") else {
            print("[KaiPlugin] ❌ Missing required parameter: userId")
            call.reject("Missing required parameter: userId")
            return
        }
        
        let authToken = call.getString("authToken")
        let backendUrl = getBackendUrl(call)
        let urlStr = "\(backendUrl)/api/kai/preferences/\(userId)"
        print("[KaiPlugin] 🌐 URL: \(urlStr)")
        
        guard let url = URL(string: urlStr) else {
            call.reject("Invalid URL: \(urlStr)")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        urlSession.dataTask(with: request) { data, response, error in
            if let error = error {
                let errorMsg = "Network error: \(error.localizedDescription) | backendUrl: \(backendUrl)"
                print("[KaiPlugin] ❌ \(errorMsg)")
                call.reject(errorMsg)
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                print("[KaiPlugin] 📡 Response status: \(httpResponse.statusCode)")
                if !(200...299).contains(httpResponse.statusCode) {
                    let bodyStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "no body"
                    let truncatedBody = bodyStr.count > 200 ? String(bodyStr.prefix(200)) + "..." : bodyStr
                    let errorMsg = "HTTP Error \(httpResponse.statusCode) | backendUrl: \(backendUrl) | body: \(truncatedBody)"
                    print("[KaiPlugin] ❌ \(errorMsg)")
                    call.reject(errorMsg)
                    return
                }
            }
            
            guard let data = data else {
                print("[KaiPlugin] ❌ No data received")
                call.reject("No data received")
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    print("[KaiPlugin] ✅ getPreferences success")
                    call.resolve(json)
                } else {
                    print("[KaiPlugin] ❌ Invalid response format")
                    call.reject("Invalid response format")
                }
            } catch {
                print("[KaiPlugin] ❌ JSON parsing error: \(error.localizedDescription)")
                call.reject("JSON parsing error: \(error.localizedDescription)")
            }
        }.resume()
    }

    @objc func resetPreferences(_ call: CAPPluginCall) {
        print("[KaiPlugin] 🔍 resetPreferences called")
        guard let userId = call.getString("userId"),
              let vaultOwnerToken = call.getString("vaultOwnerToken") else {
            call.reject("Missing required parameters: userId, vaultOwnerToken")
            return
        }

        let backendUrl = getBackendUrl(call)
        let urlStr = "\(backendUrl)/api/kai/preferences/\(userId)"
        print("[KaiPlugin] 🌐 URL: \(urlStr)")

        guard let url = URL(string: urlStr) else {
            call.reject("Invalid URL: \(urlStr)")
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(vaultOwnerToken)", forHTTPHeaderField: "Authorization")

        urlSession.dataTask(with: request) { data, response, error in
            if let error = error {
                call.reject("Network error: \(error.localizedDescription)")
                return
            }

            if let httpResponse = response as? HTTPURLResponse {
                if !(200...299).contains(httpResponse.statusCode) {
                    let body = data.flatMap { String(data: $0, encoding: .utf8) } ?? "no body"
                    call.reject("HTTP Error \(httpResponse.statusCode): \(body)")
                    return
                }
            }

            guard let data = data else {
                call.reject("No data received")
                return
            }

            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    call.resolve(json)
                } else {
                    call.resolve(["success": true])
                }
            } catch {
                call.reject("JSON parsing error: \(error.localizedDescription)")
            }
        }.resume()
    }
}

File: hushh-webapp/android/app/src/main/java/com/hushh/pda/plugins/Kai/KaiPlugin.kt
package com.hushh.pda.plugins.Kai

import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.json.JSONArray
import java.io.IOException

/**
 * Kai Plugin - Android Implementation
 * 
 * Native plugin for Agent Kai stock analysis.
 * Makes HTTP calls to backend from native code.
 */

@CapacitorPlugin(name = "Kai")
class KaiPlugin : Plugin() {
    
    private fun getBackendUrl(): String {
        // Read from environment/config, default to localhost
        val backendUrl = bridge.config.getString("backendUrl") 
            ?: System.getenv("NEXT_PUBLIC_BACKEND_URL")
            ?: "http://localhost:8000"
        
        // Android emulator: convert localhost to 10.0.2.2
        return if (backendUrl.contains("localhost")) {
            backendUrl.replace("localhost", "10.0.2.2")
        } else {
            backendUrl
        }
    }
    
    @PluginMethod
    fun grantConsent(call: PluginCall) {
        val userId = call.getString("userId") ?: run {
            call.reject("Missing userId")
            return
        }
        
        val scopesArray = call.getArray("scopes") ?: run {
            call.reject("Missing scopes")
            return
        }
        
        val authToken = call.getString("authToken")
        val backendUrl = getBackendUrl()
        val url = "$backendUrl/api/kai/consent/grant"
        
        val json = JSONObject().apply {
            put("user_id", userId)
            put("scopes", scopesArray)
        }
        
        val client = OkHttpClient()
        val body = json.toString().toRequestBody("application/json".toMediaType())
        
        val requestBuilder = Request.Builder().url(url).post(body)
        
        if (authToken != null) {
            requestBuilder.addHeader("Authorization", "Bearer $authToken")
        }
        
        val request = requestBuilder.build()
        val pluginCall = call // Rename to avoid shadowing in callback
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: okhttp3.Call, e: IOException) {
                pluginCall.reject("Network error: ${e.message}")
            }
            
            override fun onResponse(call: okhttp3.Call, response: Response) {
                val responseBody = response.body?.string()
                
                if (!response.isSuccessful || responseBody == null) {
                    pluginCall.reject("Request failed: ${response.code}")
                    return
                }
                
                try {
                    val result = JSObject(responseBody)
                    pluginCall.resolve(result)
                } catch (e: Exception) {
                    pluginCall.reject("JSON parsing error: ${e.message}")
                }
            }
        })
    }
    
    @PluginMethod
    fun analyze(call: PluginCall) {
        val userId = call.getString("userId") ?: run {
            call.reject("Missing userId")
            return
        }
        val ticker = call.getString("ticker") ?: run {
            call.reject("Missing ticker")
            return
        }
        val consentToken = call.getString("consentToken") ?: run {
            call.reject("Missing consentToken")
            return
        }
        val riskProfile = call.getString("riskProfile") ?: run {
            call.reject("Missing riskProfile")
            return
        }
        val processingMode = call.getString("processingMode") ?: run {
            call.reject("Missing processingMode")
            return
        }
        
        val authToken = call.getString("authToken")
        val backendUrl = getBackendUrl()
        val url = "$backendUrl/api/kai/analyze"
        
        val json = JSONObject().apply {
            put("user_id", userId)
            put("ticker", ticker)
            put("consent_token", consentToken)
            put("risk_profile", riskProfile)
            put("processing_mode", processingMode)
        }
        
        val client = OkHttpClient()
        val body = json.toString().toRequestBody("application/json".toMediaType())
        
        val requestBuilder = Request.Builder().url(url).post(body)
        
        if (authToken != null) {
            requestBuilder.addHeader("Authorization", "Bearer $authToken")
        }
        
        val request = requestBuilder.build()
        val pluginCall = call
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: okhttp3.Call, e: IOException) {
                pluginCall.reject("Network error: ${e.message}")
            }
            
            override fun onResponse(call: okhttp3.Call, response: Response) {
                val responseBody = response.body?.string()
                
                if (!response.isSuccessful || responseBody == null) {
                    pluginCall.reject("Request failed: ${response.code}")
                    return
                }
                
                try {
                    // Return full response directly, matching web plugin
                    val result = JSObject(responseBody)
                    pluginCall.resolve(result)
                } catch (e: Exception) {
                    pluginCall.reject("JSON parsing error: ${e.message}")
                }
            }
        })
    }
    
    @PluginMethod
    fun storePreferences(call: PluginCall) {
        val userId = call.getString("userId") ?: run {
            call.reject("Missing userId")
            return
        }

        // Canonical payload: preferences array (preferred)
        val preferencesArray = call.getArray("preferences")
        val preferencesEncrypted = call.getString("preferencesEncrypted") // legacy
        if (preferencesArray == null && preferencesEncrypted == null) {
            call.reject("Missing preferences payload")
            return
        }
        
        val authToken = call.getString("authToken")
        val backendUrl = getBackendUrl()
        val url = "$backendUrl/api/kai/preferences/store"
        
        val json = JSONObject().apply {
            put("user_id", userId)
            when {
                preferencesArray != null -> put("preferences", preferencesArray)
                else -> {
                    // Legacy: parse stringified JSON array
                    val parsed = JSONArray(preferencesEncrypted)
                    put("preferences", parsed)
                }
            }
        }
        
        val client = OkHttpClient()
        val body = json.toString().toRequestBody("application/json".toMediaType())
        
        val requestBuilder = Request.Builder().url(url).post(body)
        
        if (authToken != null) {
            requestBuilder.addHeader("Authorization", "Bearer $authToken")
        }
        
        val request = requestBuilder.build()
        val pluginCall = call
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: okhttp3.Call, e: IOException) {
                pluginCall.reject("Network error: ${e.message}")
            }
            
            override fun onResponse(call: okhttp3.Call, response: Response) {
                val responseBody = response.body?.string()
                
                if (!response.isSuccessful || responseBody == null) {
                    pluginCall.reject("Request failed: ${response.code}")
                    return
                }
                
                try {
                    val result = JSObject(responseBody)
                    pluginCall.resolve(result)
                } catch (e: Exception) {
                    pluginCall.reject("JSON parsing error: ${e.message}")
                }
            }
        })
    }
    
    @PluginMethod
    fun getPreferences(call: PluginCall) {
        val userId = call.getString("userId") ?: run {
            call.reject("Missing userId")
            return
        }
        
        val authToken = call.getString("authToken")
        val backendUrl = getBackendUrl()
        val url = "$backendUrl/api/kai/preferences/$userId"
        
        android.util.Log.d("KaiPlugin", "🔍 getPreferences called for userId: $userId")
        android.util.Log.d("KaiPlugin", "🌐 URL: $url")
        
        val client = OkHttpClient()
        
        val requestBuilder = Request.Builder().url(url).get()
        
        if (authToken != null) {
            requestBuilder.addHeader("Authorization", "Bearer $authToken")
            android.util.Log.d("KaiPlugin", "🔑 Auth token added")
        }
        
        val request = requestBuilder.build()
        val pluginCall = call
        
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: okhttp3.Call, e: IOException) {
                android.util.Log.e("KaiPlugin", "❌ Network error getting preferences: ${e.message}")
                pluginCall.reject("Network error: ${e.message}")
            }
            
            override fun onResponse(call: okhttp3.Call, response: Response) {
                val responseBody = response.body?.string()
                
                android.util.Log.d("KaiPlugin", "📡 Response code: ${response.code}")
                android.util.Log.d("KaiPlugin", "📦 Response body: ${responseBody?.take(200)}")
                
                if (!response.isSuccessful || responseBody == null) {
                    android.util.Log.e("KaiPlugin", "❌ Failed to get preferences: HTTP ${response.code}")
                    pluginCall.reject("Request failed: ${response.code}")
                    return
                }
                
                try {
                    val result = JSObject(responseBody)
                    android.util.Log.d("KaiPlugin", "✅ Preferences retrieved successfully")
                    pluginCall.resolve(result)
                } catch (e: Exception) {
                    android.util.Log.e("KaiPlugin", "❌ JSON parsing error: ${e.message}")
                    pluginCall.reject("JSON parsing error: ${e.message}")
                }
            }
        })
    }

    @PluginMethod
    fun resetPreferences(call: PluginCall) {
        val userId = call.getString("userId") ?: run {
            call.reject("Missing userId")
            return
        }
        val vaultOwnerToken = call.getString("vaultOwnerToken") ?: run {
            call.reject("Missing vaultOwnerToken")
            return
        }

        val backendUrl = getBackendUrl()
        val url = "$backendUrl/api/kai/preferences/$userId"

        val client = OkHttpClient()
        val request = Request.Builder()
            .url(url)
            .delete()
            .addHeader("Authorization", "Bearer $vaultOwnerToken")
            .build()

        val pluginCall = call
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: okhttp3.Call, e: IOException) {
                pluginCall.reject("Network error: ${e.message}")
            }

            override fun onResponse(call: okhttp3.Call, response: Response) {
                val responseBody = response.body?.string()
                if (!response.isSuccessful) {
                    pluginCall.reject("Request failed: ${response.code}")
                    return
                }
                try {
                    val result = if (responseBody.isNullOrBlank()) {
                        JSObject().put("success", true)
                    } else {
                        JSObject(responseBody)
                    }
                    pluginCall.resolve(result)
                } catch (e: Exception) {
                    pluginCall.resolve(JSObject().put("success", true))
                }
            }
        })
    }
}
